frontend-react/src/components/AuthElement.tsx

```diff
@@ -0,0 +1,58 @@
+import { useNavigate } from "react-router-dom";
+import React, { useEffect, useMemo } from "react";
+
+import { useSessionContext } from "../contexts/SessionContext";
+import { MemberType } from "../hooks/UseOktaMemberships";
+import { CheckFeatureFlag, FeatureFlagName } from "../pages/misc/FeatureFlags";
+
+interface AuthElementProps {
+    element: JSX.Element;
+    requiredUserType?: MemberType | MemberType[];
+    requiredFeatureFlag?: FeatureFlagName;
+}
+/** Wrap elements with required feature flag and/or user type checks */
+export const AuthElement = ({
+    element,
+    requiredUserType,
+    requiredFeatureFlag,
+}: AuthElementProps): React.ReactElement => {
+    // Router's new navigation hook for redirecting
+    const navigate = useNavigate();
+    const { oktaToken, activeMembership } = useSessionContext();
+    // This memberType won't require an undefined fallback when used like the one
+    // from useSessionContext
+    const memberType = useMemo(
+        () => activeMembership?.memberType || MemberType.NON_STAND,
+        [activeMembership]
+    );
+    // Dynamically authorize user from single or multiple allowed user types
+    const authorizeMemberType = useMemo(() => {
+        if (memberType === MemberType.PRIME_ADMIN) return true; // Admin authentication always
+        return requiredUserType instanceof Array
+            ? requiredUserType.includes(memberType)
+            : requiredUserType === memberType;
+    }, [requiredUserType, memberType]);
+    // All the checks before returning the route
+    useEffect(() => {
+        if (!oktaToken?.accessToken || !activeMembership) {
+            navigate("/login");
```

Comment:
I think the refresh bug we're seeing is due to token and membership being briefly undefined when the app starts up. Not sure what the best way to handle that might be. We could be really dumb and track the initial load here or within useOktaMemberships, and wait until the second render to do anything? Ok I see why you want to tackle this in a separate ticket. Maybe worth just commenting here until we figure it out?

Reply:
> We could be really dumb and track the initial load

So far as I've tried, it's a non-solution for this behavior. Because our application doesn't require users to refresh for any expected behavior, I think it's safe if this is a bug left for follow-up work.